common/map.h

```diff
@@ -0,0 +1,462 @@
+  // produce a new value when no existing value is found.
+  //
+  // TODO: The `;` formatting below appears to be bugs in clang-format with
+  // concepts that should be filed upstream.
+  template <typename LookupKeyT, typename ValueCallbackT>
+  auto Insert(LookupKeyT lookup_key, ValueCallbackT value_cb) -> InsertKVResult
+    requires(
+        !std::same_as<ValueT, ValueCallbackT> &&
+        std::convertible_to<decltype(std::declval<ValueCallbackT>()()), ValueT>)
+  ;
+
+  // Insert a key into the map and call the provided callback to allow in-place
+  // construction of both the key and value when needed.
+  template <typename LookupKeyT, typename InsertCallbackT>
+  auto Insert(LookupKeyT lookup_key, InsertCallbackT insert_cb)
+      -> InsertKVResult
+    requires(!std::same_as<ValueT, InsertCallbackT> &&
+             std::invocable<InsertCallbackT, LookupKeyT, void*, void*>);
+
+  // Similar to insert, but an existing value is replaced with the provided one.
+  template <typename LookupKeyT>
+  auto Update(LookupKeyT lookup_key, ValueT new_v) -> InsertKVResult;
+
+  // Similar to insert, but an existing value is replaced with the result of the
+  // callback.
+  template <typename LookupKeyT, typename ValueCallbackT>
+  auto Update(LookupKeyT lookup_key, ValueCallbackT value_cb) -> InsertKVResult
+    requires(
+        !std::same_as<ValueT, ValueCallbackT> &&
+        std::convertible_to<decltype(std::declval<ValueCallbackT>()()), ValueT>)
+  ;
+
+  // Similar to import, but with a distinct callback for updating an existing
+  // key/value pair as opposed to inserting a new one.
+  template <typename LookupKeyT, typename InsertCallbackT,
+            typename UpdateCallbackT>
+  auto Update(LookupKeyT lookup_key, InsertCallbackT insert_cb,
+              UpdateCallbackT update_cb) -> InsertKVResult
+    requires(!std::same_as<ValueT, InsertCallbackT> &&
+             std::invocable<InsertCallbackT, LookupKeyT, void*, void*> &&
+             std::invocable<UpdateCallbackT, KeyT&, ValueT&>);
+
+  // Erase a key from the map.
+  template <typename LookupKeyT>
+  auto Erase(LookupKeyT lookup_key) -> bool;
+
+  // Clear all key/value pairs from the map but leave the underlying hashtable
+  // allocated and in place.
+  void Clear();
+
+  // Count the probed keys. This routine is purely informational and for use in
+  // benchmarking or logging of performance anomalies. It's returns have no
+  // semantic guarantee at all.
+  auto CountProbedKeys() const -> ssize_t {
+    return ViewT(*this).CountProbedKeys();
+  }
+
+ protected:
+  using ImplT::ImplT;
+};
+
+// A data structure mapping from key to value.
+//
+// This map also supports SSO or small size optimization. The provided
+// `SmallSize` type parameter indicates the small size buffer embedded. The
+// default is zero, which always allocates a heap buffer on construction. When
+// non-zero, must be a multiple of the `MaxGroupSize` of the underlying
```

Comment:
Added the current size and I see no reason for it to grow or change really... I also mention that this is checked at compile time and so can't go wrong silently.

But do you think it is worth exposing a constant in the API so that users can programmatically compute sizes?

Or I could change the API to automatically select the next greater valid small size. I didn't do this initially because it seems unfortunate for users to request a small size of `1` or `2` and silently get `16`. I added a comment about this. But happy to revisit the API design if you think there is a better one, I don't feel super strongly here.

Reply:
I'm fine with the new text, and the policy that prioritizes readers of the code over writers.